nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -23,63 +23,93 @@ contract WorkLock {
     event Bid(address indexed sender, uint256 depositedETH);
     event Claimed(address indexed sender, uint256 claimedTokens);
     event Refund(address indexed sender, uint256 refundETH, uint256 completedWork);
-    event Burnt(address indexed sender, uint256 value);
     event Canceled(address indexed sender, uint256 value);
+    event BiddersChecked(address indexed sender, uint256 startIndex, uint256 endIndex);
+    event ForceRefund(address indexed sender, address indexed bidder, uint256 refundETH);
+    event CompensationWithdrawn(address indexed sender, uint256 value);
 
     struct WorkInfo {
         uint256 depositedETH;
         uint256 completedWork;
         bool claimed;
+        uint128 index;
     }
 
     NuCypherToken public token;
     StakingEscrow public escrow;
+    address public creator;
 
     uint256 public startBidDate;
     uint256 public endBidDate;
+    uint256 public endCancellationDate;
 
     /*
     * @dev WorkLock calculations:
-    * depositRate = tokenSupply / ethSupply
-    * claimedTokens = depositedETH * depositRate
-    * refundRate = depositRate * SLOWING_REFUND / boostingRefund
+    * bid = minBid + bonusETHPart
+    * bonusTokenSupply = tokenSupply - bidders.length * minBid
+    * bonusDepositRate = bonusTokenSupply / bonusETHSupply
+    * claimedTokens = minAllowableLockedTokens + bonusETHPart * bonusDepositRate
+    * bonusRefundRate = bonusDepositRate * SLOWING_REFUND / boostingRefund
     * refundETH = completedWork / refundRate
     */
     uint256 public boostingRefund;
     uint16 public constant SLOWING_REFUND = 100;
     uint256 private constant MAX_ETH_SUPPLY = 2e10 ether;
 
+    uint256 public minAllowedBid;
     uint256 public tokenSupply;
-    uint256 public ethSupply;
-    uint256 public unclaimedTokens;
+    uint256 public bonusETHSupply;
     uint16 public stakingPeriods;
     mapping(address => WorkInfo) public workInfo;
+    mapping(address => uint256) public compensation;
+
+    address[] public bidders;
+    // if value == bidders.length then WorkLock is fully checked
+    uint256 public nextBidderToCheck;
+    // copy from the escrow contract
+    uint256 public maxAllowableLockedTokens;
+    uint256 public minAllowableLockedTokens;
+
+    /**
+    * @dev Checks timestamp regarding cancellation window
+    */
+    modifier afterCancellationWindow()
+    {
+        require(block.timestamp >= endCancellationDate,
+            "Operation is allowed when cancellation phase is over");
```

Comment:
```suggestion
            "Operation not allowed during cancellation phase ");
```

Reply:
Hmm, we have also time before cancelation phase so it's why here "after"  but maybe it's ok to change to "during"